sources/lib/structures/typed.ml

```diff
@@ -187,178 +203,440 @@ and 'a tdecl =
+  match a1, a2 with
+    TAtrue, TAtrue
+  | TAfalse, TAfalse -> true
+  | TAeq l1, TAeq l2
+  | TAdistinct l1, TAdistinct l2
+  | TAneq l1, TAneq l2
+  | TAle l1, TAle l2
+  | TAlt l1, TAlt l2 ->
+    eq_list
+      (fun t1 t2 -> eq_tterm t1.c t2.c)
+      l1
+      l2
+  | TApred (t1, b1), TApred (t2, b2) ->
+    (b1 && (not b2) || (not b1) && b2) &&
+    eq_tterm t1.c t2.c
+  | TTisConstr (t1, s1), TTisConstr (t2, s2) ->
+    Hstring.equal s1 s2 && eq_tterm t1.c t2.c
+  | _,_ -> false
+
+and eq_quant_form (q1 : 'a quant_form) (q2 : 'a quant_form) : bool =
+  let stylist =
+    eq_list
+      (fun (s1,t1) (s2, t2) -> Symbols.equal s1 s2 && Ty.equal t1 t2)
+  in
+  stylist q1.qf_bvars q2.qf_bvars &&
+  stylist q1.qf_upvars q2.qf_upvars &&
+  eq_list (fun f1 f2 -> eq_tform f1.c f2.c) q1.qf_hyp q2.qf_hyp &&
+  eq_tform q1.qf_form.c q2.qf_form.c &&
+  eq_list
+    (fun (tlist1,b1) (tlist2,b2) ->
+       (b1 && (not b2) || (not b1) && b2) &&
+       eq_list
+         (fun t1 t2 -> eq_tterm t1.c t2.c)
+         tlist1
+         tlist2)
+    q1.qf_triggers
+    q2.qf_triggers
+
+and eq_tform (f1 : 'a tform) (f2 : 'a tform) : bool =
+  match f1, f2 with
+  | TFatom a1, TFatom a2 -> eq_tatom a1.c a2.c
+
+  | TFop (op1, l1), TFop (op2, l2) ->
+    eq_oplogic op1 op2 &&
+    eq_list
+      (fun f1 f2 -> eq_tform f1.c f2.c)
+      l1
+      l2
+
+  | TFforall q1, TFforall q2
+  | TFexists q1, TFexists q2 ->
+    eq_quant_form q1 q2
+
+  | TFlet (stlist1, slklist1,f1), TFlet (stlist2, slklist2,f2) ->
+    let stylist =
+      eq_list
+        (fun (s1,t1) (s2, t2) -> Symbols.equal s1 s2 && Ty.equal t1 t2)
+    in
+    stylist stlist1 stlist2 &&
+    eq_list
+      (fun (s1,lf1) (s2, lf2) -> Symbols.equal s1 s2 && eq_tlet_kind lf1 lf2)
+      slklist1 slklist2
+    &&
+    eq_tform f1.c f2.c
+
+  | TFnamed (s1, f1), TFnamed (s2, f2) ->
+    Hstring.equal s1 s2 && eq_tform f1.c f2.c
+
+  | TFmatch (t1, pfl1), TFmatch (t2, pfl2) ->
+    eq_tterm t1.c t2.c &&
+    eq_list
+      (fun (p1, f1) (p2, f2) -> eq_pattern p1 p2 && eq_tform f1.c f2.c)
+      pfl1
+      pfl2
+
+  | _,_ -> false
+
+and eq_tlet_kind (k1 : 'a tlet_kind) (k2 : 'a tlet_kind) : bool =
+  match k1, k2 with
+    TletTerm t1, TletTerm t2 -> eq_tterm t1.c t2.c
+  | TletForm f1, TletForm f2 -> eq_tform f1.c f2.c
+  | _,_ -> false
+
 (*****)
 
 let string_of_op = function
   | OPand -> "and"
   | OPor -> "or"
   | OPimp -> "->"
   | OPiff -> "<->"
+  | OPxor -> "xor"
+  | OPif -> "ite"
   | _ -> assert false
 
+type 'annot annot_printer = Format.formatter -> 'annot -> unit
+type ('a,'annot) annoted_printer =
+  Format.formatter -> ('a,'annot) annoted -> unit
+
+let (no_print : _ annot_printer) = fun fmt _ -> fprintf fmt ""
+let (int_print : int annot_printer) = fun fmt -> fprintf fmt ".%i"
+
+let print_annot
+    (annot : 'annot annot_printer)
+    (print : ('a,'annot) annoted_printer)
+    (fmt : Format.formatter) (t : ('a, 'annot) annoted) =
+  fprintf fmt "%a%a" print t annot t.annot
```

Comment:
From what I understand of this code, an `annoted_printer` actually does not print annotations, contrary to what seems intuitive when reading the name.
In any case, why do you need annotation printing here ?

Reply:
Not sure to follow what you mean. An `annoted_printer` prints `annot`s independently of their annotations. It is the function `annot` that prints the annotation. For clarity, I rename the argument `annot` of `print_annot` into `pp_annot`.
There is as much interest in printing annotation as there is in printing terms. When the simplifyer was working with terms, it was useful (for debugging) to check the differences between structurally equal terms with different annotations. Now that the simplifyer is no longer working on `typed.ml`types, the printing utils are not used anymore.